### PR TITLE
fix(demo): move the try/catch again

### PR DESCRIPTION
**Pull request by raquelmsmith**
## Problem

Apparently when a `transaction.atomic()` fails, it does so kind of silently. The recommendation is to not catch the exception inside the transaction because then it may commit bad data to the database. See: https://docs.djangoproject.com/en/4.1/topics/db/transactions/#:~:text=Avoid%20catching%20exceptions%20inside%20atomic!

Sooooooo here's another shot at debugging this!

## Changes

Moved the try/except block to outside the transaction, but inside another transaction, in the hopes that the error will get raised.

I'm copy/pasting code from the internet, particularly as it involves the DatabaseError. @Twixes your eyes would be appreciated here.

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/api/team.py [hunk 1]
```diff
@@ -135,19 +135,23 @@ def validate(self, attrs: Any) -> Any:
...
+        with transaction.atomic():
+            try:
+                with transaction.atomic():
+                    team = Team.objects.create_with_data(**validated_data, organization=organization)
+                    request.user.current_team = team
+                    request.user.save()
+                    if validated_data.get("is_demo", False):
+                        MatrixManager(HedgeboxMatrix(), use_pre_save=True).run_on_team(team, request.user)
+            except DatabaseError as db_err:
+                capture_exception()
```
Comment (pauldambra) [inline posthog/api/team.py:150]: fly-by sticking my nose in with no context 😊 
```suggestion
                capture_exception(db_err)
```

I think you need to pass the exception into capture....

File: posthog/api/team.py [hunk 2]
```diff
@@ -135,19 +135,23 @@ def validate(self, attrs: Any) -> Any:
...
+                raise db_err
+            except Exception as e:  # TODO: Remove this after 2022-12-07, the except is just temporary for debugging
+                capture_exception()
```
Comment (pauldambra) [inline posthog/api/team.py:153]:
```suggestion
                capture_exception(e)
```

File: posthog/api/team.py (hunk 1 above)
Comment (neilkakkar) [reply posthog/api/team.py]: I was shocked when I first learned this, but sentry's SDK captures exceptions with or without passing them in https://docs.sentry.io/platforms/python/usage/

File: posthog/api/team.py [hunk 3]
```diff
@@ -135,19 +135,23 @@ def validate(self, attrs: Any) -> Any:
...
+        organization = self.context["view"].organization  # Use the org we used to validate permissions
+        with transaction.atomic():
+            try:
+                with transaction.atomic():
+                    team = Team.objects.create_with_data(**validated_data, organization=organization)
+                    request.user.current_team = team
+                    request.user.save()
+                    if validated_data.get("is_demo", False):
+                        MatrixManager(HedgeboxMatrix(), use_pre_save=True).run_on_team(team, request.user)
+            except DatabaseError as db_err:
```
Comment (neilkakkar) [inline posthog/api/team.py:149]: this bit doesn't really do anything, is a duplicate of below?

**Comment by neilkakkar** [comment]
I think you can just have a single transaction here and catch it inside. The recommendation you link to is against *handling* that exception, but since you're raising it again, there's no issues.

**Comment by neilkakkar** [comment]
Although, I'm pretty surprised by the transaction failing silently 🤔 - that seems like unexpected behaviour to me / not what the docs say should happen. How'd you manage to figure out that it's failing silently?

**Comment by raquelmsmith** [comment]
> Although, I'm pretty surprised by the transaction failing silently 🤔 - that seems like unexpected behaviour to me / not what the docs say should happen. How'd you manage to figure out that it's failing silently?

The docs say that it just rolls it back and doesn't complete it. I get a 500 but no error in Sentry and no helpful info in the 500.

**Comment by raquelmsmith** [comment]
I'm out of my depth here so if someone who has more python/django experience wants to help me out with this I would be extremely appreciative and also buy you a beer.

File: posthog/api/team.py (hunk 3 above)
Comment (neilkakkar) [reply posthog/api/team.py:149]: Yep, `DatabaseError` inherits from base `Exception`, so the latter catches the former as well

**Comment by neilkakkar** [comment]
A link to the Sentry error please? This is weird 🤔

**Comment by neilkakkar** [comment]
lol, just re-read what you sent 🤦 , duh no sentry error.

Umm, how do I reproduce?

aaand I guess doesn't happen locally?

**Comment by raquelmsmith** [comment]
Correct, does not happen locally. 

1. Log into some posthog account in prod (don't use the PostHog posthog acct, use your own test one). 
2. Go to /ingestion. 
3. In console, `posthog.feature_flags.override({'onboarding-v2-experiment': 'test'})`
4. Reload.
5. Click the "I just want to try with demo data." button.

It is expected to take a minute to copy the demo data over. But then it just fails on prod, but not locally.

**Comment by neilkakkar** [comment]
I'm 70% sure the problem is `gunicorn` timeouts: (or some other timeout I couldn't find)

https://github.com/PostHog/posthog/blob/master/gunicorn.config.py#L22 which is set to 15 seconds right now.

We don't run `gunicorn` when running locally. One quick way to confirm would be trying to start the backend locally with gunicorn.

What's happening is that there's nothing going wrong in the transaction, it's just slow, and gets cut off when it exceeds the timeout, and none of our codepaths unfortunately tell that to sentry yet.

---

One open q I'm not sure about: insights endpoint can take a lot longer to return, did we special case that, or maybe I'm totally off the mark 😅

**Comment by raquelmsmith** [comment]
OK that's helpful! Thank you! What do you recommend to make this work?

**Comment by neilkakkar** [comment]
Assuming it is indeed the reason, we can make this call asynchronous, where the API returns immediately, and polls while it waits for the operation to finish in the background.

For an example, we do exactly this with cohort calculations.

See: https://github.com/PostHog/posthog/blob/master/posthog/tasks/calculate_cohort.py#L37  -> on an `update` request to a cohort, we fire this async task which does what's needed, and the UI polls until it gets a positive response https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/cohorts/cohortEditLogic.ts#L244

File: posthog/api/team.py (hunk 1 above)
Comment (pauldambra) [reply posthog/api/team.py]: OMG 🤯 

That's too much magic, I don't like 🧙

Comment (neilkakkar) [reply posthog/api/team.py]: I wholeheartedly agree

**Comment by neilkakkar** [comment]
How'd you get along with this?

**Comment by neilkakkar** [comment]
How'd you get along with this?

**Comment by raquelmsmith** [comment]
@kappa90 and I tried moving it into a task but kept running into issues with the demo script. So I passed it over to @Twixes and he's done some debugging. It's still not working so I'm hoping to look back into it soon, unless @Twixes can!

**Comment by raquelmsmith** [comment]
Closing in favor of #13228

**Pull request closed without merging**